keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2144 @@
+  should therefore plan the maintenance in advance and ensure that pods are gracefully removed
+  before attempting to shut down or restart the machine.
+- The kubelet has no way of detecting ongoing maintenance if the node is restarted in the meantime.
+- Graceful termination of static pods during a shutdown is not possible today. It is also not
+  possible to keep them terminated after the machine has been restarted if the node is still under
+  maintenance. 
+
+
+### Motivation Summary
+
+To sum up. Some applications solve the disruption problem by introducing validating admission
+webhooks. This has some drawbacks. The webhooks are not easily discoverable by cluster admins. And
+they can block evictions for other applications if they are misconfigured or misbehave. The
+eviction API is not intended to be extensible in this way. The webhook approach is therefore not
+recommended.
+
+Some drainers solve the node drain by depending on the kubectl logic, or by extending/rewriting it
+with additional rules and logic.
+
+As seen in the experience reports and GitHub issues, some admins solve their problems by simply
+ignoring PDBs which can cause unnecessary disruptions or data loss. Some solve this by playing
+with the application deployment, but have to understand that the application supports this.
+
+Kubelet's graceful node shutdown feature is a best-effort solution for unplanned shutdowns, but
+it is not sufficient to ensure application and data safety.
+
+### Goals
+- Introduce NodeMaintenance API.
+- Introduce a node maintenance controller that creates evacuations
+- Deprecate kubectl drain in favor of NodeMaintenance.
+- Implement NodeMaintenanceAwareKubelet feature to implement a lifecycle for Static pods during a
+  maintenance.
+- Implement NodeMaintenanceAwareDaemonSet feature to prevent the scheduling of DaemonSet pods on
+  nodes during a maintenance.
+
+### Non-Goals
+- Introduce a node maintenance period, nodeDrainTimeout (similar to [cluster-api](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/control-plane)
+  nodeDrainTimeout) or a TTL optional field as an upper bound on the duration of node maintenance.
+  Then the node maintenance would be garbage collected and the node made schedulable again.
+
+## Proposal
+
+Most of these issues stem from the lack of a standardized way of detecting a start of the node
+drain. This KEP proposes the introduction of a NodeMaintenance object that would signal an intent
+to gracefully remove pods from given nodes. The intent will be implemented by the newly proposed
+[Evacuation API KEP](https://github.com/kubernetes/enhancements/issues/4563), which ensures graceful pod removal or migration, an ability to measure
+the progress and a fallback to eviction if progress is lost. The NodeMaintenance implementation
+should also utilize existing node's `.spec.unschedulable` field, which prevents new pods from being
```

Comment:
Wouldn't this result in the same issues as what we have seen currently between the Graceful Node Shutdown manager and the DaemonSet controller? See:

- https://github.com/kubernetes/kubernetes/issues/122912

DaemonSet controller, through the nature of its design, will ignore most of the taints and Node statuses.

Reply:
Please look at the [DaemonSet Controller](https://github.com/kubernetes/enhancements/blob/a15b78d3b4f8bc1b4a4e5b1508e4ad733e64c4fd/keps/sig-apps/4212-declarative-node-maintenance/README.md#daemonset-controller) section. It will have to observe the NodeMaintenance API to adhere to the maintenance.